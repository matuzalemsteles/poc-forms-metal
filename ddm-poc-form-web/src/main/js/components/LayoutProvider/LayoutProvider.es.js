import Component, { Fragment } from 'metal-jsx';
import {Config} from 'metal-state';

import { LayoutSupport } from 'ddm-poc-form-js-components/Layout/index.js';

/**
 * LayoutProvider listens to your children's events to 
 * control the `context` and make manipulations.
 * @extends Component
 */
class LayoutProvider extends Component {
    static PROPS = {
        /**
         * @default undefined
         * @instance
         * @memberof LayoutProvider
         * @type {?(array|undefined)}
         */
        context: Config.array(),

        /**
         * @default undefined
         * @instance
         * @memberof LayoutProvider
         * @type {?(array|undefined)}
         */
        spritemap: Config.string()
    }

    static STATE = {
        /**
         * @default undefined
         * @instance
         * @memberof LayoutProvider
         * @type {?array}
         */
        context: Config.array(),

        /**
         * @default undefined
         * @instance
         * @memberof LayoutProvider
         * @type {?object}
         */
        fieldFocus: Config.shapeOf({
            indexColumn: Config.oneOfType([
                Config.bool().value(false),
                Config.number()
            ]).required(),
            indexPage: Config.number().required(),
            indexRow: Config.number().required(),
            type: Config.string().required(),
        }).value({}),

        /**
         * @default add
         * @instance
         * @memberof LayoutProvider
         * @type {?string}
         */
        mode: Config.oneOf(['add', 'edit']).value('add'),
    }

    /**
	 * @inheritDoc
	 */
    constructor(props, context) {
        super(props, context);

        this.state.context = props.context;
    }

    /**
     * @param {!Object} data
     * @private
     */
    _handleFieldClicked(data) {
        this.setState({
            fieldFocus: data,
            mode: 'edit'
        });
    }

    /**
     * @param {!Object}
     * @private
     */
    _handleFieldAdd({ target, fieldProperties, data }) {
        const { spritemap } = this.props;
        const { context } = this.state;
        const { indexRow, indexPage, indexColumn } = target;

        fieldProperties = Object.assign({}, fieldProperties, {spritemap});

        let newContext = null;

        if (target.indexColumn === false) {
            let newRow = LayoutSupport.implAddRow(12, [fieldProperties]);
            newContext = LayoutSupport.addRow(context, indexRow, indexPage, newRow);
        } else {
            newContext = LayoutSupport.addFieldToColumn(context, indexPage, indexRow, indexColumn, fieldProperties);
        }

        this.setState({
            context: newContext,
            fieldFocus: {
                indexColumn,
                indexPage,
                indexRow,
                type: fieldProperties.type,
            },
            mode: 'edit'
        });
    }

    /**
     * @param {!Object}
     * @private
     */
    _handleFieldDelete({ indexRow, indexPage, indexColumn }) {
        const { context } = this.state;
        let newContext = LayoutSupport.removeFields(context, indexPage, indexRow, indexColumn);
        newContext = this._cleanRowEmpty(newContext, { indexRow, indexPage, indexColumn });

        this.setState({
            context: newContext,
            fieldFocus: {}
        });
    }

    /**
     * @param {!Object}
     * @private
     */
    _handleFieldEdit({value, key}) {
        const { context, fieldFocus } = this.state;
        const { indexColumn, indexPage, indexRow } = fieldFocus;
        const fieldSelected = LayoutSupport.getColumn(context, indexPage, indexRow, indexColumn);

        const implPropertiesField = {
            [key]: value
        }

        const newField = Object.assign({}, fieldSelected[0], implPropertiesField);

        const newContext = LayoutSupport.changeFieldsFromColumn(context, indexPage, indexRow, indexColumn, [newField]);

        this.setState({
            context: this.state.context
        });
    }

    /**
     * @param {!Object}
     * @private
     */
    _handleFieldMove({ target, source }) {
        const { context } = this.state;
        const fieldSourceToMove = this._getFieldSourceToMove(context, source);

        let newContext = LayoutSupport.removeFields(context, source.indexPage, source.indexRow, source.indexColumn);

        if (target.indexColumn === false) {
            newContext = this._cleanRowEmpty(newContext, source);
            newContext = this._addFieldToRow(newContext, target, fieldSourceToMove);
        } else {
            newContext = this._addFieldToColumn(newContext, target, fieldSourceToMove);
            newContext = this._cleanRowEmpty(newContext, source);
        }

        this.setState({
            context: newContext,
            fieldFocus: {}
        });

        newContext = null;
    }

    /**
     * @param {!Array} context
     * @param {!Object} source
     * @private
     * @return {Object}
     */
    _cleanRowEmpty(context, source) {
        const { indexRow, indexPage, indexColumn } = source;

        if (LayoutSupport.hasFieldsRow(context, indexPage, indexRow)) {
            return LayoutSupport.removeRow(context, indexPage, indexRow);
        }

        return context;
    }

    /**
     * @param {!Array} context
     * @param {!Object} field
     * @param {!Object} target
     * @private
     * @return {Object}
     */
    _addFieldToRow(context, target, field) {
        const { indexRow, indexPage } = target;
        const newRow = LayoutSupport.implAddRow(12, field);
        return LayoutSupport.addRow(context, indexRow, indexPage, newRow);
    }

    /**
     * @param {!Array} context
     * @param {!Object} field
     * @param {!Object} target
     * @private
     * @return {Object}
     */
    _addFieldToColumn(context, target, field) {
        const { indexRow, indexPage, indexColumn } = target;
        return LayoutSupport.addFields(context, indexPage, indexRow, indexColumn, field);
    }

    /**
     * @param {!Array} context
     * @param {!Object} source
     * @private
     * @return {Object}
     */
    _getFieldSourceToMove(context, source) {
        const { indexRow, indexPage, indexColumn } = source;
        return LayoutSupport.getColumn(context, indexPage, indexRow, indexColumn);
    }

    render() {
        const { children } = this.props;
        const { fieldFocus, context, mode } = this.state;

        if (children.length) {
            const Child = children[0];

            const events = {
                fieldAdd: this._handleFieldAdd.bind(this),
                fieldEdit: this._handleFieldEdit.bind(this),
                fieldClicked: this._handleFieldClicked.bind(this),
                fieldDelete: this._handleFieldDelete.bind(this),
                fieldMove: this._handleFieldMove.bind(this),
            };

            Object.assign(Child.props, {...this.otherProps(), events, context, fieldFocus, mode});

            return Child;
        }

        return;
    }
}

export default LayoutProvider;