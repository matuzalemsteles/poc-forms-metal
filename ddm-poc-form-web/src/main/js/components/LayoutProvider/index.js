import LayoutProvider from './LayoutProvider.es';

export default LayoutProvider;