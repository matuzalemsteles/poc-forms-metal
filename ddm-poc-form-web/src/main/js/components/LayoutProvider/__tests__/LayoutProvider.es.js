import {JSXComponent} from 'metal-jsx';
import Context from './__mock__/mockContext.es';
import LayoutProvider from '../LayoutProvider.es';

let component;
let context = null;
let spritemap = 'icons.svg';

class Child extends JSXComponent {
    render() {
        return <div />
    }
}

class Parent extends JSXComponent {
    render() {
        return(
            <LayoutProvider context={context} spritemap={spritemap} ref="provider">
                <Child ref="child" />
            </LayoutProvider>
        );
    }
}

describe('LayoutProvider', () => {
    beforeEach(() => {
        context = JSON.parse(JSON.stringify(Context));

        jest.useFakeTimers();
    });

    afterEach(() => {
        if (component) {
            component.dispose();
        }

        context = null;
    });

    it('should receive context through PROPS and move to the internal state', () => {
        component = new LayoutProvider({
            context,
        });

        expect(component.state.context).toEqual(context);
    });

    it('should attach the events to the child component', () => {
        component = new Parent();

        const { provider } = component.refs;
        
        expect(provider.props.children[0].props.events).toMatchObject({
            fieldAdd: expect.any(Function),
            fieldClicked: expect.any(Function),
            fieldDelete: expect.any(Function),
            fieldEdit: expect.any(Function),
            fieldMove: expect.any(Function)
        });
    });

    it('should pass to the child component the context of the internal state', () => {
        component = new Parent();

        const { provider, child } = component.refs;

        expect(child.props.context).toEqual(provider.state.context);
    });

    it('should pass to the child component the fieldFocus', () => {
        component = new Parent();

        const { provider, child } = component.refs;

        const fieldFocus = {
            indexColumn: 0,
            indexPage: 0,
            indexRow: 0,
            type: 'radio'
        };

        provider.setState({
            fieldFocus,
        });

        jest.runAllTimers();

        expect(child.props.fieldFocus).toEqual(provider.state.fieldFocus);
    });

    it('should pass to the child component the mode', () => {
        component = new Parent();

        const { provider, child } = component.refs;

        provider.setState({
            mode: 'edit'
        });

        jest.runAllTimers();

        expect(child.props.mode).toEqual(provider.state.mode);
    });

    describe('Field events', () => {
        describe('fieldMove', () => {
            it('should listen to the fieldMove event and move the field to the row in context', () => {
                component = new Parent();
    
                const { provider, child } = component.refs;
                const mockEvent = {
                    target: {
                        indexPage: 0,
                        indexRow: 0,
                        indexColumn: false
                    },
                    source: {
                        indexColumn: 0,
                        indexRow: 1,
                        indexPage: 0
                    }
                };
    
                child.emit('fieldMove', mockEvent);
    
                jest.runAllTimers();
    
                expect(provider.state.context).toMatchSnapshot();
                expect(child.props.context).toEqual(provider.state.context);
            });
    
            it('should listen to the fieldMove event and move the field to the column in context', () => {
                component = new Parent();
    
                const { provider, child } = component.refs;
                const mockEvent = {
                    target: {
                        indexColumn: 1,
                        indexPage: 0,
                        indexRow: 0
                    },
                    source: {
                        indexColumn: 1,
                        indexPage: 0,
                        indexRow: 1
                    }
                };
    
                child.emit('fieldMove', mockEvent);
    
                jest.runAllTimers();
    
                expect(provider.state.context).toMatchSnapshot();
                expect(child.props.context).toEqual(provider.state.context);
            });
    
            it('should move the field to the column in context and remove the row if there are no fields', () => {
                component = new Parent();
    
                const { provider, child } = component.refs;
                const mockEvent = {
                    target: {
                        indexColumn: 1,
                        indexRow: 0,
                        indexPage: 0
                    },
                    source: {
                        indexColumn: 0,
                        indexPage: 0,
                        indexRow: 2
                    }
                };
    
                child.emit('fieldMove', mockEvent);
    
                jest.runAllTimers();
    
                expect(provider.state.context).toMatchSnapshot();
                expect(child.props.context).toEqual(provider.state.context);
            });
    
            it('should move the field to the row in context and remove the row if there are no fields', () => {
                component = new Parent();
    
                const { provider, child } = component.refs;
                const mockEvent = {
                    target: {
                        indexColumn: false,
                        indexRow: 0,
                        indexPage: 0
                    },
                    source: {
                        indexColumn: 0,
                        indexPage: 0,
                        indexRow: 0
                    }
                };
    
                child.emit('fieldMove', mockEvent);
    
                jest.runAllTimers();
    
                expect(provider.state.context).toMatchSnapshot();
                expect(child.props.context).toEqual(provider.state.context);
            });
        });

        describe('fieldAdd', () => {
            it('should listen the fieldAdd event and add the field in the column to the context', () => {
                component = new Parent();
    
                const { provider, child } = component.refs;
                const mockEvent = {
                    target: {
                        indexRow: 0,
                        indexPage: 0,
                        indexColumn: 1
                    },
                    fieldProperties: {
                        type: 'text',
                    }
                };
    
                child.emit('fieldAdd', mockEvent);
    
                jest.runAllTimers();
    
                expect(provider.state.context).toMatchSnapshot();
                expect(child.props.context).toEqual(provider.state.context);
            });
    
            it('should listen the fieldAdd event and add the field in the row to the context', () => {
                component = new Parent();
    
                const { provider, child } = component.refs;
                const mockEvent = {
                    target: {
                        indexRow: 0,
                        indexPage: 0,
                        indexColumn: false
                    },
                    fieldProperties: {
                        type: 'text',
                    }
                };
    
                child.emit('fieldAdd', mockEvent);
    
                jest.runAllTimers();
    
                expect(provider.state.context).toMatchSnapshot();
                expect(child.props.context).toEqual(provider.state.context);
            });
    
            it('should update the fieldFocus with the location of the new field when adding to the context', () => {
                component = new Parent();
    
                const { provider, child } = component.refs;
                const mockEvent = {
                    target: {
                        indexRow: 1,
                        indexPage: 0,
                        indexColumn: 2
                    },
                    fieldProperties: {
                        type: 'text',
                    }
                };
    
                child.emit('fieldAdd', mockEvent);
    
                const fieldFocus = {
                    ...mockEvent.target,
                    type: mockEvent.fieldProperties.type
                };
    
                jest.runAllTimers();
    
                expect(provider.state.fieldFocus).toEqual(fieldFocus);
                expect(child.props.fieldFocus).toEqual(fieldFocus);
                expect(provider.state.mode).toBe('edit');
            });
        });

        describe('fieldDelete', () => {
            it('should listen the fieldDelete event and delete the field in the column to the context', () => {
                component = new Parent();
    
                const { provider, child } = component.refs;
                const mockEvent = {
                    indexRow: 1,
                    indexPage: 0,
                    indexColumn: 0
                };
                
                child.emit('fieldDelete', mockEvent);
    
                jest.runAllTimers();
    
                expect(provider.state.context).toMatchSnapshot();
                expect(child.props.context).toEqual(provider.state.context);
            });
        });

        describe('fieldEdit', () => {
            it('should listen the fieldEdit event and edit the field to the context', () => {
                component = new Parent();
    
                const { provider, child } = component.refs;
                const mockEvent = {
                    value: 'Foo',
                    key: 'label'
                };
                const mockFieldFocus = {
                    indexColumn: 0,
                    indexPage: 0,
                    indexRow: 1
                };
    
                child.emit('fieldClicked', mockFieldFocus);
                child.emit('fieldEdit', mockEvent);
    
                jest.runAllTimers();
    
                expect(
                    provider
                        .state
                        .context[mockFieldFocus.indexPage]
                        .rows[mockFieldFocus.indexRow]
                        .columns[mockFieldFocus.indexColumn]
                        .fields[0][mockEvent.key]
                ).toBe(mockEvent.value);
                expect(provider.state.context).toMatchSnapshot();
                expect(child.props.context).toEqual(provider.state.context);
            });
        });

        describe('fieldClicked', () => {
            it('should listen the fieldClicked event and change the state of the fieldFocus to the data receive', () => {
                component = new Parent();
        
                const { provider, child } = component.refs;
                const mockEvent = {
                    indexColumn: 0,
                    indexPage: 0,
                    indexRow: 0,
                    mode: 'edit'
                };
        
                child.emit('fieldClicked', mockEvent);
        
                expect(provider.props.children[0].props.events).toMatchObject({
                    fieldClicked: expect.any(Function)
                });
                expect(provider.state.fieldFocus).toEqual(mockEvent);
            });
        });
    });
});