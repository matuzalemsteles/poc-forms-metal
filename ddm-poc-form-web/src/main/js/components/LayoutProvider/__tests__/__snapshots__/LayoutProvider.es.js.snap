// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LayoutProvider Field events fieldAdd should listen the fieldAdd event and add the field in the column to the context 1`] = `
Array [
  Object {
    "rows": Array [
      Object {
        "columns": Array [
          Object {
            "fields": Array [
              Object {
                "label": "Radio",
                "type": "radio",
              },
            ],
            "size": 3,
          },
          Object {
            "fields": Array [
              Object {
                "spritemap": "icons.svg",
                "type": "text",
              },
            ],
            "size": 9,
          },
        ],
      },
      Object {
        "columns": Array [
          Object {
            "fields": Array [
              Object {
                "label": "Text",
                "required": true,
                "type": "text",
              },
            ],
            "size": 4,
          },
          Object {
            "fields": Array [
              Object {
                "items": Array [
                  Object {
                    "name": "Foo",
                  },
                  Object {
                    "name": "Bar",
                  },
                ],
                "label": "Select",
                "required": true,
                "type": "select",
              },
            ],
            "size": 6,
          },
          Object {
            "fields": Array [],
            "size": 2,
          },
        ],
      },
      Object {
        "columns": Array [
          Object {
            "fields": Array [
              Object {
                "label": "Second row",
                "required": true,
                "type": "text",
              },
            ],
            "size": 12,
          },
        ],
      },
    ],
    "title": "Untitled name",
  },
]
`;

exports[`LayoutProvider Field events fieldAdd should listen the fieldAdd event and add the field in the row to the context 1`] = `
Array [
  Object {
    "rows": Array [
      Object {
        "columns": Array [
          Object {
            "fields": Array [
              Object {
                "spritemap": "icons.svg",
                "type": "text",
              },
            ],
            "size": 12,
          },
        ],
      },
      Object {
        "columns": Array [
          Object {
            "fields": Array [
              Object {
                "label": "Radio",
                "type": "radio",
              },
            ],
            "size": 3,
          },
          Object {
            "fields": Array [],
            "size": 9,
          },
        ],
      },
      Object {
        "columns": Array [
          Object {
            "fields": Array [
              Object {
                "label": "Text",
                "required": true,
                "type": "text",
              },
            ],
            "size": 4,
          },
          Object {
            "fields": Array [
              Object {
                "items": Array [
                  Object {
                    "name": "Foo",
                  },
                  Object {
                    "name": "Bar",
                  },
                ],
                "label": "Select",
                "required": true,
                "type": "select",
              },
            ],
            "size": 6,
          },
          Object {
            "fields": Array [],
            "size": 2,
          },
        ],
      },
      Object {
        "columns": Array [
          Object {
            "fields": Array [
              Object {
                "label": "Second row",
                "required": true,
                "type": "text",
              },
            ],
            "size": 12,
          },
        ],
      },
    ],
    "title": "Untitled name",
  },
]
`;

exports[`LayoutProvider Field events fieldDelete should listen the fieldDelete event and delete the field in the column to the context 1`] = `
Array [
  Object {
    "rows": Array [
      Object {
        "columns": Array [
          Object {
            "fields": Array [
              Object {
                "label": "Radio",
                "type": "radio",
              },
            ],
            "size": 3,
          },
          Object {
            "fields": Array [],
            "size": 9,
          },
        ],
      },
      Object {
        "columns": Array [
          Object {
            "fields": Array [
              Object {
                "label": "Second row",
                "required": true,
                "type": "text",
              },
            ],
            "size": 12,
          },
        ],
      },
    ],
    "title": "Untitled name",
  },
]
`;

exports[`LayoutProvider Field events fieldEdit should listen the fieldEdit event and edit the field to the context 1`] = `
Array [
  Object {
    "rows": Array [
      Object {
        "columns": Array [
          Object {
            "fields": Array [
              Object {
                "label": "Radio",
                "type": "radio",
              },
            ],
            "size": 3,
          },
          Object {
            "fields": Array [],
            "size": 9,
          },
        ],
      },
      Object {
        "columns": Array [
          Object {
            "fields": Array [
              Object {
                "label": "Foo",
                "required": true,
                "type": "text",
              },
            ],
            "size": 4,
          },
          Object {
            "fields": Array [
              Object {
                "items": Array [
                  Object {
                    "name": "Foo",
                  },
                  Object {
                    "name": "Bar",
                  },
                ],
                "label": "Select",
                "required": true,
                "type": "select",
              },
            ],
            "size": 6,
          },
          Object {
            "fields": Array [],
            "size": 2,
          },
        ],
      },
      Object {
        "columns": Array [
          Object {
            "fields": Array [
              Object {
                "label": "Second row",
                "required": true,
                "type": "text",
              },
            ],
            "size": 12,
          },
        ],
      },
    ],
    "title": "Untitled name",
  },
]
`;

exports[`LayoutProvider Field events fieldMove should listen to the fieldMove event and move the field to the column in context 1`] = `
Array [
  Object {
    "rows": Array [
      Object {
        "columns": Array [
          Object {
            "fields": Array [
              Object {
                "label": "Radio",
                "type": "radio",
              },
            ],
            "size": 3,
          },
          Object {
            "fields": Array [
              Object {
                "items": Array [
                  Object {
                    "name": "Foo",
                  },
                  Object {
                    "name": "Bar",
                  },
                ],
                "label": "Select",
                "required": true,
                "type": "select",
              },
            ],
            "size": 9,
          },
        ],
      },
      Object {
        "columns": Array [
          Object {
            "fields": Array [
              Object {
                "label": "Second row",
                "required": true,
                "type": "text",
              },
            ],
            "size": 12,
          },
        ],
      },
    ],
    "title": "Untitled name",
  },
]
`;

exports[`LayoutProvider Field events fieldMove should listen to the fieldMove event and move the field to the row in context 1`] = `
Array [
  Object {
    "rows": Array [
      Object {
        "columns": Array [
          Object {
            "fields": Array [
              Object {
                "label": "Text",
                "required": true,
                "type": "text",
              },
            ],
            "size": 12,
          },
        ],
      },
      Object {
        "columns": Array [
          Object {
            "fields": Array [
              Object {
                "label": "Radio",
                "type": "radio",
              },
            ],
            "size": 3,
          },
          Object {
            "fields": Array [],
            "size": 9,
          },
        ],
      },
      Object {
        "columns": Array [
          Object {
            "fields": Array [
              Object {
                "label": "Second row",
                "required": true,
                "type": "text",
              },
            ],
            "size": 12,
          },
        ],
      },
    ],
    "title": "Untitled name",
  },
]
`;

exports[`LayoutProvider Field events fieldMove should move the field to the column in context and remove the row if there are no fields 1`] = `
Array [
  Object {
    "rows": Array [
      Object {
        "columns": Array [
          Object {
            "fields": Array [
              Object {
                "label": "Radio",
                "type": "radio",
              },
            ],
            "size": 3,
          },
          Object {
            "fields": Array [
              Object {
                "label": "Second row",
                "required": true,
                "type": "text",
              },
            ],
            "size": 9,
          },
        ],
      },
      Object {
        "columns": Array [
          Object {
            "fields": Array [
              Object {
                "label": "Text",
                "required": true,
                "type": "text",
              },
            ],
            "size": 4,
          },
          Object {
            "fields": Array [
              Object {
                "items": Array [
                  Object {
                    "name": "Foo",
                  },
                  Object {
                    "name": "Bar",
                  },
                ],
                "label": "Select",
                "required": true,
                "type": "select",
              },
            ],
            "size": 6,
          },
          Object {
            "fields": Array [],
            "size": 2,
          },
        ],
      },
      Object {
        "columns": Array [
          Object {
            "fields": Array [],
            "size": 12,
          },
        ],
      },
    ],
    "title": "Untitled name",
  },
]
`;

exports[`LayoutProvider Field events fieldMove should move the field to the row in context and remove the row if there are no fields 1`] = `
Array [
  Object {
    "rows": Array [
      Object {
        "columns": Array [
          Object {
            "fields": Array [
              Object {
                "label": "Radio",
                "type": "radio",
              },
            ],
            "size": 12,
          },
        ],
      },
      Object {
        "columns": Array [
          Object {
            "fields": Array [],
            "size": 3,
          },
          Object {
            "fields": Array [],
            "size": 9,
          },
        ],
      },
      Object {
        "columns": Array [
          Object {
            "fields": Array [
              Object {
                "label": "Text",
                "required": true,
                "type": "text",
              },
            ],
            "size": 4,
          },
          Object {
            "fields": Array [
              Object {
                "items": Array [
                  Object {
                    "name": "Foo",
                  },
                  Object {
                    "name": "Bar",
                  },
                ],
                "label": "Select",
                "required": true,
                "type": "select",
              },
            ],
            "size": 6,
          },
          Object {
            "fields": Array [],
            "size": 2,
          },
        ],
      },
      Object {
        "columns": Array [
          Object {
            "fields": Array [
              Object {
                "label": "Second row",
                "required": true,
                "type": "text",
              },
            ],
            "size": 12,
          },
        ],
      },
    ],
    "title": "Untitled name",
  },
]
`;
